src/lib-smtp/fuzz-smtp-server.c

```diff
@@ -0,0 +1,96 @@
+/* Copyright (c) 2020 Catena cyber */
+
+#include "lib.h"
+#include "str.h"
+#include "array.h"
+#include "hostpid.h"
+#include "ioloop.h"
+#include "istream.h"
+#include "ostream.h"
+#include "time-util.h"
+#include "sleep.h"
+#include "connection.h"
+#include "test-common.h"
+#include "smtp-address.h"
+#include "smtp-reply-parser.h"
+#include "smtp-server.h"
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <fcntl.h>
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size);
+
+static struct ioloop *ioloop_fuzz;
+static int initialized = 0;
+int fdo[2];
+int fdi[2];
+
+struct server_connection {
+    void *context;
+};
+
+
+static void server_connection_destroy(void *context)
+{
+    struct server_connection *sconn = (struct server_connection *)context;
+    i_free(sconn);
+}
+
+static void
+test_server_defaults(struct smtp_server_settings *smtp_set)
+{
+    /* server settings */
+    i_zero(smtp_set);
+    smtp_set->max_client_idle_time_msecs = 500;
+    smtp_set->max_pipelined_commands = 16;
+    smtp_set->auth_optional = TRUE;
+    smtp_set->debug = 0;
+}
+
+
+static void test_server_continue(void *dummy) {
+    //instead of simple io_loop_stop so as to free input io
+    io_loop_stop_delayed(ioloop_fuzz);
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+    struct smtp_server_connection *conn;
+    struct server_connection *sconn;
+    struct timeout *to;
+    struct smtp_server_settings smtp_server_set;
+    static struct smtp_server_callbacks server_callbacks;
+    static struct smtp_server *smtp_server = NULL;
+
+    if (initialized == 0) {
+        lib_init();
+        hostpid_init();
+        initialized = 1;
+    }
+    test_server_defaults(&smtp_server_set);
+    i_zero(&server_callbacks);
+    server_callbacks.conn_destroy = server_connection_destroy;
+    ioloop_fuzz = io_loop_create();
+    to = timeout_add_short(0, test_server_continue, NULL);
+    smtp_server = smtp_server_init(&smtp_server_set);
+
+    /* build input as istream out of data and size */
+    sconn = i_new(struct server_connection, 1);
+    pipe(fdi);
+    pipe(fdo);
+    write(fdi[1], data, size);
+    conn = smtp_server_connection_create(smtp_server, fdi[0], fdo[1],
+                                         NULL, 0, FALSE, NULL, &server_callbacks, sconn);
```

Comment:
I wonder if it works to use smtp_server_connection_create_from_streams() instead? That would avoid using the pipe and would be a bit more reliable, especially with large inputs where write to pipe might not write everything.

```
input = i_stream_create_from_data(data, size);
output = o_stream_create_fd(dev_null_fd, IO_BLOCK_SIZE);
conn = smtp_server_connection_create_from_streams(...)
```

Reply:
Oh well, looks like it doesn't work. It really wants fds to exist.

Reply:
Indeed, I has tried with `smtp_server_connection_create_from_streams` without success as fds were required.

I found that `pipe` are the most socket-like object without needing another process, but you are welcome if you find a better way to do it :-)